### ISSUE TITLE
python.d.plugin: high cpu usage, database moves into the future

### ISSUE THREAD

**Issue by ktsaou** [firehol/netdata#2328]
@l2isbad check this:

![image](https://user-images.githubusercontent.com/2662304/27119419-84f548e8-50e8-11e7-8034-8f72d3fe0dbb.png)

The CPU utilization of python.d.plugin spikes to 100%.

netdata also logs this:

2017-06-14 09:59:23: netdata INFO : RRD database for chart 'cpu.cpu0_cpuidle' on host 'igw2' is 21.26085 secs in the past. Adjusting it to current time.
2017-06-14 09:59:23: netdata INFO : RRD database for chart 'cpu.cpu1_cpuidle' on host 'igw2' is 21.26113 secs in the past. Adjusting it to current time.
2017-06-14 09:59:23: netdata INFO : RRD database for chart 'cpu.cpu2_cpuidle' on host 'igw2' is 21.26139 secs in the past. Adjusting it to current time.
2017-06-14 09:59:23: netdata INFO : RRD database for chart 'cpu.cpu3_cpuidle' on host 'igw2' is 21.26218 secs in the past. Adjusting it to current time.
2017-06-14 09:59:23: netdata INFO : RRD database for chart 'netdata.plugin_pythond_cpuidle' on host 'igw2' is 21.26238 secs in the past. Adjusting it to current time.
2017-06-14 09:59:24: netdata INFO : RRD database for chart 'web_log_sq_viva.squid_response_statuses' on host 'igw2' is 1.43910 secs in the future. Adjusting it to current time.
2017-06-14 09:59:24: netdata INFO : RRD database for chart 'web_log_sq_viva.squid_response_codes' on host 'igw2' is 1.43911 secs in the future. Adjusting it to current time.
2017-06-14 09:59:24: netdata INFO : RRD database for chart 'web_log_sq_viva.squid_detailed_response_codes' on host 'igw2' is 1.43917 secs in the future. Adjusting it to current time.

This current python.d.plugin is unstable.

Any ideas how to debug this?

**Comment by ktsaou** [comment]
Also, all python.d.plugin charts show gaps:

![image](https://user-images.githubusercontent.com/2662304/27119534-e0541ade-50e8-11e7-85f9-e819a12bde05.png)

It does not seem web_log related. Even this shows gaps:

![image](https://user-images.githubusercontent.com/2662304/27119600-26e2f5ba-50e9-11e7-85bb-7f32299ef8f7.png)

and these:

![image](https://user-images.githubusercontent.com/2662304/27119635-46d0a9bc-50e9-11e7-8603-6417fc0d0920.png)

This servers runs python.d.plugin with web_log (squid and apache), cpuidle and postfix.

**Comment by ktsaou** [comment]
Also it seems that it does something CPU intensive once per minute.

This chart shows that yesterdays merges were not the problematic ones (the gap is the daily updater job - CPU spikes existed before yesterdays merges):

![image](https://user-images.githubusercontent.com/2662304/27120158-2e9fabde-50eb-11e7-9dfa-cdf5954065a7.png)

**Comment by l2isbad** [comment]
Well, does the problem exist right after python.d.plugin/netdata restart?

If so i suggest to run plugins in debug mode one by one and see which of them is the problem. Or it's not the plugin but python.d.plugin.

**Comment by ktsaou** [comment]
I disabled all except web_log: problem exists.
I disabled all except squid (native, not web_log): problem does not exist.
I disabled web_log and enabled all others: problem does not exist.

So, it is in web_log.

**Comment by ktsaou** [comment]
First half without web_log, second half with web_log:

![image](https://user-images.githubusercontent.com/2662304/27120996-3e6ce15a-50ee-11e7-9139-5b5106edb922.png)

**Comment by ktsaou** [comment]
I think this is related to the number of log lines in the log. Squid on this machine logs 100 entries per second.

**Comment by l2isbad** [comment]
> 100 entries per second

100 is not too much

i checked with 200+ per second on office pc (model name: Intel(R) Pentium(R) CPU G2020 2.90GHz
)
![screenshot_20170614_165618](https://user-images.githubusercontent.com/22274335/27121492-90f4998e-5122-11e7-9a3a-fb1c4727ac2b.png)
![screenshot_20170614_165642](https://user-images.githubusercontent.com/22274335/27121497-9887a880-5122-11e7-9441-2f9dff706237.png)
![screenshot_20170614_165658](https://user-images.githubusercontent.com/22274335/27121506-9f59b2d4-5122-11e7-97a9-f53e3fd4e093.png)

**Comment by l2isbad** [comment]
I think it's the LogService problem. web_log is just parse the data which is fast.

**Comment by l2isbad** [comment]
1000+ per secod on the same pc
![screenshot_20170614_170359](https://user-images.githubusercontent.com/22274335/27121763-8d69d5d0-5123-11e7-816f-6f7f1d44b645.png)
![screenshot_20170614_170413](https://user-images.githubusercontent.com/22274335/27121774-94e78078-5123-11e7-9f10-3b4ff2cf4fca.png)
![screenshot_20170614_170424](https://user-images.githubusercontent.com/22274335/27121788-9eb185fe-5123-11e7-92f9-ca23154622f1.png)

**Comment by ktsaou** [comment]
The config I have is like this:

```
squid_log:
  name: 'squid'
  type: 'squid'
  path: '/var/log/squid3/access.log'
  filter:
    exclude: '.*.(microsoft.com|azure.com|office365.com|office.com|AAA.gr|XXX.com|YYY.com)'

squid_microsoft_log:
  name: 'sq_ms'
  type: 'squid'
  path: '/var/log/squid3/access.log'
  filter:
    include: '.*.(microsoft.com|azure.com|office365.com|office.com)'

squid_AAA_log:
  name: 'sq_AAA'
  type: 'squid'
  path: '/var/log/squid3/access.log'
  filter:
    include: '.*.AAA.gr'

squid_XXX_log:
  name: 'sq_XXX'
  type: 'squid'
  path: '/var/log/squid3/access.log'
  filter:
    include: '.*.(XXX.com|YYY.com)'
```

Could it be related to filteting?

**Comment by ktsaou** [comment]
Yes, it is. I disabled all filtering and left only one job for squid and the problem is gone.
So, it is filtering related.

**Comment by l2isbad** [comment]
confirm that. checked it myself

**Comment by l2isbad** [comment]
@ktsaou 

```bash
[/var/log/squid3]wc -l test
2176 test
```

```python
from itertools import ifilterfalse as filterfalse
import re
regex = re.compile(r'.*.(microsoft.com|azure.com|office365.com|office.com|AAA.gr|XXX.com|YYY.com)')
#regex = re.compile(r'(?:microsoft.com|azure.com|office365.com|office.com|AAA.gr|XXX.com|YYY.com)')
def test(v):
    return not regex.search(v)

f = open('test')
f = f.readlines()
for i in filterfalse(test, f):
    pass
```

```bash
[/var/log/squid3]time python 2.py 

real    0m0.842s
user    0m0.824s
sys     0m0.016s
```

```python
from itertools import ifilterfalse as filterfalse
import re
# !!! 
regex = re.compile(r'(?:microsoft.com|azure.com|office365.com|office.com|AAA.gr|XXX.com|YYY.com)')
# !!!
def test(v):
    return not regex.search(v)

f = open('test')
f = f.readlines()
for i in filterfalse(test, f):
    pass
```
```bash
[/var/log/squid3]time python 2.py 

real    0m0.023s
user    0m0.020s
sys     0m0.000s
```

**Comment by l2isbad** [comment]
And all `(X|Y|Z)` should be `(?:X|Y|Z)`

**Comment by ktsaou** [comment]
nice finding!

**Comment by ktsaou** [comment]
What is strange though, is how this affected the other python modules.... Why cpuidle and sensors got gaps?

**Comment by ktsaou** [comment]
The regex changes did this:

![image](https://user-images.githubusercontent.com/2662304/27129292-58a8aa6a-510a-11e7-850a-d028d2173773.png)

Sure, this did a significant improvement.

**Comment by ktsaou** [comment]
@l2isbad many-many thanks for this!

This is still somewhat strange:

> What is strange though, is how this affected the other python modules.... Why cpuidle and sensors got gaps?

I understand the the only way for this happen, is if the whole python process is single threaded. So, the jobs are not actually separate linux threads, but an async-I/O trick (like node.js does).

**Comment by Ferroin** [comment]
On the specific note of other charts being blocked, it's probably a side-effect of Python's Global Interpreter Lock.  In essence, A multi-threaded Python application can only have one thread running in each process at a time, because there's some internal state in the interpreter itself that needs to be global for threads which can't be concurrently accessed.  Usually, this isn't noticeable because most stuff that uses threads is I/O bound and written to avoid polling, but when you end up with something that needs to poll, or are doing CPU or memory bound operations, you end up with issues like this.

**Comment by ktsaou** [comment]
thanks!

I am trying to figure out a way for netdata to detect this and rase an alarm or something... Any ideas?

**Comment by Ferroin** [comment]
My first thought is to add some way to match on known metrics not being received for extended periods of time.

If I was coding it myself, I'd probably figure out which plugins are not likely to block due to external interference (that is, ones which don't depend on some external service running (so cpuidle and the proc plugin would qualify, but the nginx, postfix, and mysql ones wouldn't), track which of those we've seen data from, and then set things up to send an alert (or warn in the logs, or whatever else) when we don't see data from such plugins for more than some configurable threshold.  That would have the advantage that it would also allow you to send an alert when the system is so heavily loaded that netdata can't keep up (which based on my own experience, usually means that something is wrong, even if it's not an issue in netdata itself).

I'm not certain that there's any way it can be reliably restricted to specific plugins though, but in the Python case this could be mitigated by using a process pool instead of a thread pool, although that may impact performance somewhat (at a minimum, it will increase the number of context-switches during collection).

**Comment by ktsaou** [comment]
hm... This works already. This how I found the issue. squid (not the web_log squid, but the native squid plugin) was raising alarms about not being collected.

But, even in my case, I got really confused. I thought recents commits made the plugin unstable. The right way to alarm about it should be something that would alarm about web_log high CPU usage. I will try to restore the wrong patterns to find out what the netdata plugins report (the last few charts on the dashboard that track the timings of the plugins).
